src/models/losses.py

```diff
@@ -4,6 +4,327 @@
 import torch
 from torch import nn
 
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-locals
+
+
+def generator_loss(
+    logits: dict[str, dict[str, torch.Tensor]],
+    inception_encoder: Any,
+    fake_imgs: torch.Tensor,
+    real_labels: torch.Tensor,
+    words_emb: torch.Tensor,
+    sent_emb: torch.Tensor,
+    match_labels: torch.Tensor,
+    cap_lens: torch.Tensor,
+    class_ids: torch.Tensor,
+    vgg_encoder: Any,
+    real_imgs: torch.Tensor,
+    device: Any,
+    const_dict: dict[Any, Any],
+) -> Any:
+    """Calculate the loss for the generator.
+
+    Args:
+        logits: Dictionary with fake/real and word-level/uncond/cond logits
+
+        inception_encoder: The inception encoder model.
+
+        fake_imgs: The fake images generated by the generator. [shape: (batch_size, 3, 256, 256)]
+
+        real_labels: Label for "real" image as predicted by discriminator,
+        this is a tensor of ones. [shape: (batch_size, 1)].
+
+        words_emb: The embeddings for all the words in the captions.
+        shape: (batch_size, embedding_size, max_caption_length)
+
+        sent_emb: The embeddings for the sentences.
+        shape: (batch_size, embedding_size)
+
+        match_labels: Tensor of shape: (batch_size, 1).
+        This is of the form torch.tensor([0, 1, 2, ..., batch-1])
+
+        cap_lens: The length of the 'actual' captions in the batch [without padding]
+        shape: (batch_size, 1)
+
+        class_ids: The class ids for the instance. shape: (batch_size, 1)
+
+        vgg_encoder: The VGG encoder model.
```

Comment:
vgg encoder is used for perceptual loss calculation where we also pass fake images to the encoder. passing it here as features of fake image from VGG encoder is only used for perceptual loss, felt it would abstract the loss more compared to using in training loop. let me know your thoughts.

Reply:
> felt it would abstract the loss

Which is exactly what we want. Think of default pytorch losses, they don'y take networks as inputs, they take outputs and what they should be. Losses shouldn't be aware of the flow of system, they're pure mathematical functions. There's a reason why we make loss functions a separate functions and don't calculate that entirely in the training loop and that reason is that it helps **reduce coupling** between net inference and net evaluation

Reply:
i see, makes sense. i have fixed this too.